### PR TITLE
feat: Add bb interface implementation

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

This PR adds the barretenberg implementation of noir js backend interface as a new package 

## Problem\*

Resolves 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/noir_js_backend_barretenberg/src/serialize.ts
Comment (TomAFrench) [inline tooling/noir_js_backend_barretenberg/src/serialize.ts:4]:
```suggestion
// Converts bytecode from a base64 string to a Uint8Array
```

File: tooling/noir_js_backend_barretenberg/src/base64_decode.ts
Comment (kevaundray) [inline tooling/noir_js_backend_barretenberg/src/base64_decode.ts:3]: We want to move this into noir-lang/types along with serialize, so that both noir-js and this package can access it

File: tooling/noir_js_backend_barretenberg/package.json
Comment (kevaundray) [inline tooling/noir_js_backend_barretenberg/package.json:21]: This is not great, might be better to patch the artefacts with a package.json file in the future

File: tooling/noir_js_backend_barretenberg/src/index.ts
Comment (signorecello) [inline tooling/noir_js_backend_barretenberg/src/index.ts:65]: Just a smaller nit...

as suggested in slack, I wouldn't `instantiante` these dynamically, as we're adding some overhead to every proof generation... I'd add these in the `Noir` API, together with the other init functions `initACVM` and `initAbi`

this will make the app have a bit of an overhead at the start, but from then on every proof will _apparently_ be much faster

**Review by signorecello — CHANGES_REQUESTED** [review]

**Review by TomAFrench — DISMISSED** [review]
LGTM. @signorecello can you unblock this PR so we can merge the others?

**Review by TomAFrench — DISMISSED** [review]

**Review by signorecello — APPROVED** [review]
LGTM 👍

File: tooling/noir_js_backend_barretenberg/src/index.ts
Comment (kevaundray) [reply tooling/noir_js_backend_barretenberg/src/index.ts:65]: In other places we do lazy instantiation, I have added an issue here on either doing one way or the otherhttps://github.com/noir-lang/noir/issues/2910#issuecomment-1742811237

**Comment by kevaundray** [comment]
Reverted the change to instantiate backend on init for now because it forces us to make backend.instantiate public. I'll note it in the issue and come back to this once all PRs are merged. I'll ask Koby about his thinking regarding lazy initting versus not

**Comment by kevaundray** [comment]
For some reason, I cannot add Noir.Js as a required test, opened an issue [here](https://github.com/noir-lang/noir/issues/2930)

**Pull request merged**
